### DISCUSSION TITLE
How does persist middleware work with URL  and how to test it?

### DISCUSSION THREAD

**Discussion by pucheng-tan** [pmndrs/zustand#2038] [Q&A]
On this page 'Connect to state with URL': https://docs.pmnd.rs/zustand/guides/connect-to-state-with-url-hash. It shows a nice example of how to use persist middleware to store info within the URL. However, it does not describe how it works, especially I am confused about when `removeItem` would be triggered and how to test it. It would be nice if someone could provide me with some explanation or show me some code examples.

**Comment by dai-shi** [comment]
removeItem is called when clearStorage is called explicitly:
https://github.com/pmndrs/zustand/blob/266e8641b29579dc0dd0b816d28c8b5797d19e9a/src/middleware/persist.ts#L346-L348
I've never tried it, so can't help much.

**Comment by dbritto-dev** [comment]
The example is pretty clear about how works. The persist middleware works on any kind of storage (that can persist data). In this case we can use the `URL` to be more specific we can use `URLSearchParams`, through this we can `serialize` and `deserialize` whatever we want.

So we can test it, actually is pretty easy. Let's say that we have the next query string or search parameters (window.location.search)

- window.location.search: '?version=2&app=zustand&foo=bar'

We can parse them with `URLSearchParams`

- new URLSearchParams(window.location.search.slice(1))
- we can get `version`, `app` and `foo` using `URLSearchParams.get()` or `URLSearchParams.getAll()`
- We can delete `foo` using `URLSearchParams.delete()`

If we delete `foo` the `window.location.search` now is '?version=2&app=zustand'

So the best way to test it out is by checking `window.location.search` the example in the docs has all of that

**Comment by vogelino** [comment]
Hi! Great way to persist the state to the URL. Thx for that. 

I have a question related to this. I'm using it in the context of a Next.js application. It supports changing the search parameters when using `window.history.pushState` as in the demo, however, it does not trigger a rerender on components that use `useStore` when navigating using the browser back button. The components using `useSearchParams()` do rerender, as the search parameters did change, but it seems as the state doesn't change in the store. Is there any way to force the store to reflect the changes when navigating via browser history?

I hope my description was clear.

Thank you!

**Reply by DavidRamsdale** [reply]
I am having the same problem did you find a solution to this @vogelino ?

**Reply by vogelino** [reply]
> I am having the same problem did you find a solution to this @vogelino ?

Unfortunately no. I have move focus from this project so I haven't kept investigating but did not found a way to fix it.

**Reply by zavvdev** [reply]
@vogelino @DavidRamsdale 

I solved with by just listening for popstate events. Basically, my implementation is next:
1. I created Serializers for my state. For example, I use MUI DataGrid and I want to persist sortModel in url. I've added custom serialize/deserialize function so I can have something like this in URL: ?sort=field1:asc,field2:desc instead of the whole JSON string.
2. Then I created searchParamsPersistor that takes serialize/deserialize and returns Zustand interface for storage. It handles read/write to URL internally.
3. Once I have a store, I can listen to popstate even and just set my state by just deserializing it from store.

```typescript
export const createSortingStore = ({ urlStateKey }: CreateStoreArgs = {}) => {
  type SortingStore = ResetableStore & {
    model: SortModel;
    updateModel: (model: SortModel) => void;
  };

  const initialModelState: SortModel = [];

  if (!urlStateKey) {
    return create<SortingStore>()((set) => ({
      model: initialModelState,
      updateModel: (model) => set({ model }),
      reset: () => set({ model: initialModelState }),
    }));
  }

  const store = create<SortingStore>()(
    persist(
      (set) => ({
        model: initialModelState,
        updateModel: (model) => set({ model }),
        reset: () => set({ model: initialModelState }),
      }),
      {
        name: urlStateKey,
        storage: searchParamsPersistor<Pick<SortingStore, "model">>(
          (state) => SortModelSerializer.serialize(state.model),
          (serialized) => ({
            model: SortModelSerializer.deserialize(serialized),
          }),
        ),
      },
    ),
  );

  syncWithPopstate(store, urlStateKey, (serialized) => ({
    model: SortModelSerializer.deserialize(serialized),
  }));

  return store;
};

export const syncWithPopstate = <T extends object>(
  store: StoreApi<T>,
  urlStateKey: string,
  syncFn: (serialized: string) => Partial<T>,
): void => {
  window.addEventListener("popstate", () => {
    store.setState(syncFn(searchParamsUtil.getOne(urlStateKey) ?? ""));
  });
};
```

**Reply by zavvdev** [reply]
But you need to keep in mind, that when your store.setState is called from even listener, Zustand will trigger the state update, so the new URL entry will be written and you'll get a new URL record which won't allow you to go forward after you went back.
Like: 
1. Push state A -> URL: A
2. Push state B -> URL: A,B
3. Move back -> URL: A -> state setter triggers and a new URL entry pushed, so we have a duplicate record -> URL: A and forward browser button is disabled since it's the latest URL record.

So u need to handle this case in searchParamsPersistor and disallow state updates if serialized state is the same.

**Reply by zavvdev** [reply]
But to be honest, I find this solution unreliable. For example, i use MUI and store pagination, sorting, filtering models of DataGrid in Zustand + sync with URL state. When I try to sync back/forward buttons it works for sorting and filtering but for pagination it doesn't. Maybe because DataGrid generates some pagination model change events and my Zustand setter triggers and writes again in storage which creates new URL history record. So I decided to just keep unidirectional approach for now and don't support state recovery on back/forward browser navigation. It's too fragile and makes you to have 2 sources of truth for your state. So I think, if you want to support this type of functionality you need to have your URL state as a single source of truth without Zustand. Or maybe I'm wrong and someone can suggest something.

**Reply by vogelino** [reply]
Thank you @zavvdev for looking into this and sharing. I have since moved on to another project, but I remember finding this strategy to be brittle. It’s disappointing that it cannot retain or recover state when navigating back and forth. If I return to my project, I will give it a try and share my experience here. Thanks again! :)
